merlion/models/layers.py

```diff
@@ -0,0 +1,253 @@
+    """
+    Config object for a `LayeredModel`. See `LayeredModel` documentation for more details.
+    """
+
+    def __init__(self, model: Union[ModelBase, Dict], model_kwargs=None, **kwargs):
+        """
+        :param model: The model being wrapped, or a dict representing it.
+        :param model_kwargs: Keyword arguments used specifically to initialize the underlying model. Only used if
+            ``model`` is a dict. Will override keys in the ``model`` dict if specified.
+        :param kwargs: Any other keyword arguments (e.g. for initializing a base class). If ``model`` is a dict,
+            we will also try to pass these arguments when creating the actual underlying model. However, they will
+            not override arguments in either the ``model`` dict or ``model_kwargs`` dict.
+        """
+        # Model-specific kwargs override kwargs when creating the model.
+        model_kwargs = {} if model_kwargs is None else model_kwargs
+        if isinstance(model, dict):
+            model.update({k: v for k, v in kwargs.items() if k not in model and k not in model_kwargs})
+            model, extra_kwargs = ModelFactory.create(**{**model, **model_kwargs, "return_unused_kwargs": True})
+            kwargs.update(extra_kwargs)
+        self.model = model
+        super().__init__(**kwargs)
+
+        # If no model was created, reserve unused kwargs to try initializing the model with
+        if model is None:
+            extra_kwargs = {k: v for k, v in kwargs.items() if k not in self.to_dict()}
+            self.model_kwargs = {**extra_kwargs, **model_kwargs}
+
+    @property
+    def model(self):
+        """
+        The underlying model which this layer wraps.
+        """
+        return self._model
+
+    @model.setter
+    def model(self, model):
+        self._model = model
+        self.model_kwargs = {}
+
+    @property
+    def depth(self):
+        """
+        The depth of the layered model, i.e. the number of layers until we hit a base model. Base models have depth 0.
+        """
+        return 0 if self.model is None else self.model.config.depth + 1
+
+    @property
+    def base_model(self):
+        """
+        The base model at the heart of the full layered model.
+        """
+        model = self.model
+        for _ in range(self.depth - 1):
+            model = model.model
+        return model
+
+    def to_dict(self, _skipped_keys=None):
+        _skipped_keys = _skipped_keys if _skipped_keys is not None else set()
+        config_dict = super().to_dict(_skipped_keys.union({"model"}))
+        if not self.model_kwargs and "model_kwargs" in config_dict:
+            config_dict["model_kwargs"] = None
+        if "model" not in _skipped_keys:
+            if self.model is None:
+                config_dict["model"] = None
+            else:
+                config_dict["model"] = dict(name=type(self.model).__name__, **self.model.config.to_dict())
+        return config_dict
+
+    def __copy__(self):
```

Comment:
This method seems like it should be `__deepcopy__`, since we are doing a deep copy of the underlying model.

Reply:
Fair point, but the main reason for this is that `copy` and `deepcopy` are actually the same for all models and configs in the repo. We can revisit this if we choose to implement a shallow copy.